scipy/linalg/basic.py

```diff
@@ -75,34 +101,117 @@ def solve(a, b, sym_pos=False, lower=False, overwrite_a=False,
+    numpy.dot() behavior and the returned result is still 1D array.
+
+    The generic, symmetric, hermitian and positive definite solutions are
+    obtained via calling ?GESVX, ?SYSVX, ?HESVX, and ?POSVX routines of
+    LAPACK respectively.
     """
-    a1 = _asarray_validated(a, check_finite=check_finite)
-    b1 = _asarray_validated(b, check_finite=check_finite)
-    if len(a1.shape) != 2 or a1.shape[0] != a1.shape[1]:
-        raise ValueError('expected square matrix')
-    if a1.shape[0] != b1.shape[0]:
-        raise ValueError('incompatible dimensions')
-    overwrite_a = overwrite_a or _datacopied(a1, a)
-    overwrite_b = overwrite_b or _datacopied(b1, b)
-    if debug:
-        print('solve:overwrite_a=', overwrite_a)
-        print('solve:overwrite_b=', overwrite_b)
+    # Flags for 1D or nD right hand side
+    b_is_1D = False
+    b_is_ND = False
+
+    a1 = atleast_2d(_asarray_validated(a, check_finite=check_finite))
+    b1 = atleast_1d(_asarray_validated(b, check_finite=check_finite))
+    n = a1.shape[0]
+
+    if a1.shape[0] != a1.shape[1]:
+        raise ValueError('Input a needs to be a square matrix.')
+
+    if n != b1.shape[0]:
+        # Last chance to catch 1x1 scalar a and 1D b arrays
+        if not (n == 1 and b1.size != 0):
+            raise ValueError('Input b has to have same number of rows as '
+                             'input a')
+
+    # accomodate empty arrays
+    if b1.size == 0:
+        return np.asfortranarray(b1.copy())
+
+    # regularize 1D b arrays to 2D and catch nD RHS arrays
+    if b1.ndim == 1:
+        if n == 1:
+            b1 = b1[None, :]
+        else:
+            b1 = b1[:, None]
+        b_is_1D = True
+    elif b1.ndim > 2:
+        b_is_ND = True
+
+    r_or_c = complex if np.iscomplexobj(a1) else float
+
+    if assume_a in ('gen', 'sym', 'her', 'pos'):
+        _structure = assume_a
+    else:
+        raise ValueError('{} is not a recognized matrix structure'
+                         ''.format(assume_a))
+
+    # Backwards compatibility - old keyword.
     if sym_pos:
-        posv, = get_lapack_funcs(('posv',), (a1, b1))
-        c, x, info = posv(a1, b1, lower=lower,
-                          overwrite_a=overwrite_a,
-                          overwrite_b=overwrite_b)
+        assume_a = 'pos'
+
+    if _structure == 'gen':
+        gesvx = get_lapack_funcs('gesvx', (a1, b1))
+        trans_conj = 'N'
+        if transposed:
+            trans_conj = 'T' if r_or_c is float else 'H'
+        (_, _, _, _, _, _, _,
+         x, rcond, _, _, info) = gesvx(a1, b1,
+                                       trans=trans_conj,
+                                       overwrite_a=overwrite_a,
+                                       overwrite_b=overwrite_b
+                                       )
+    elif _structure == 'sym':
+        sysvx, sysvx_lw = get_lapack_funcs(('sysvx', 'sysvx_lwork'), (a1, b1))
+        lwork, _ = sysvx_lw(n, lower)
```

Comment:
This, and the others, likely should use `_compute_lwork` --- for reasons see _compute_lwork